File: frontend/src/types/listing.ts
import type { ReviewResponse } from './review';

export type ListingStatus = 'active' | 'reserved' | 'sold' | 'paused';

export interface ListingRealEstate {
  id: number;
  name: string;
  email: string;
}

export interface ListingBuyer {
  id: number;
  name: string;
  surname: string;
  email: string;
}

export interface ListingResponse {
  id: number;
  property_id: number;
  price: number;
  status: ListingStatus;
  property?: { 
    id: number;
    address: string;
    location: string;
    type: string;
    characteristics: string | null;
  };
  buyer?: ListingBuyer | null;
  real_estate?: ListingRealEstate;
  average_rating: number | null; 
  reviews: ReviewResponse[];
}

export interface ListingCreate {
  property_id: number;
  price: number;
  status: ListingStatus;
}

export interface ListingUpdate {
  price?: number;
  status?: ListingStatus;
}

export interface ListingFilters {
  location?: string;
  minPrice?: number;
  maxPrice?: number;
  propertyType?: 'house' | 'apartment';
}

File: backend/app/schema/listing.py
from pydantic import BaseModel, ConfigDict

from app.core.enums import ListingStatus
from app.schema.client import ClientResponse
from app.schema.property import PropertyResponse
from app.schema.real_estate import RealEstateResponse
from app.schema.review import ReviewResponse


class ListingCreate(BaseModel):
    property_id: int
    price: float
    status: ListingStatus = ListingStatus.ACTIVE


class ListingUpdate(BaseModel):
    property_id: int | None = None
    price: float | None = None
    status: ListingStatus | None = None


class ListingResponse(BaseModel):
    id: int
    price: float
    status: ListingStatus

    property: PropertyResponse
    real_estate: RealEstateResponse
    buyer: ClientResponse | None = None

    average_rating: float | None = None
    reviews: list[ReviewResponse] = []

    model_config = ConfigDict(from_attributes=True)
